src/Functions/FunctionsCharsetClassification.cpp

```diff
@@ -0,0 +1,247 @@
+    static constexpr size_t max_string_size = 1u << 15;
+
+    /// Default padding to read safely.
+    static constexpr size_t default_padding = 16;
+
+    /// Max codepoints to store at once. 16 is for batching usage and PODArray has this padding.
+    static constexpr size_t simultaneously_codepoints_num = default_padding + N - 1;
+
+
+    static ALWAYS_INLINE inline Float64 naiveBayes(const FrequencyHolder::EncodingMap & standard,
+        std::unordered_map<UInt16, Float64> & model,
+        Float64 max_result)
+    {
+        Float64 res = 0;
+        for (auto & el : model)
+        {
+            /// Try to find bigram in the dictionary.
+            auto it = standard.find(el.first);
+            if (it != standard.end())
+            {
+                res += el.second * log(it->getMapped());
+            } else
+            {
+                res += el.second * log(zero_frequency);
+            }
+            /// If at some step the result has become less than the current maximum, then it makes no sense to count it fully.
+            if (res < max_result)
+            {
+                return res;
+            }
+        }
+        return res;
+    }
+
+
+    static ALWAYS_INLINE size_t readCodePoints(CodePoint * code_points, const char *& pos, const char * end)
+    {
+        constexpr size_t padding_offset = default_padding - N + 1;
+        memcpy(code_points, code_points + padding_offset, roundUpToPowerOfTwoOrZero(N - 1) * sizeof(CodePoint));
+        memcpy(code_points + (N - 1), pos, default_padding * sizeof(CodePoint));
+        pos += padding_offset;
+        if (pos > end)
+            return default_padding - (pos - end);
+        return default_padding;
+    }
+
+    /// Сount how many times each bigram occurs in the text.
+    static ALWAYS_INLINE inline size_t calculateStats(
+        const char * data,
+        const size_t size,
+        size_t (*read_code_points)(CodePoint *, const char *&, const char *),
+        std::unordered_map<UInt16, Float64>& model)
+    {
+
+        const char * start = data;
+        const char * end = data + size;
+        CodePoint cp[simultaneously_codepoints_num] = {};
+        /// read_code_points returns the position of cp where it stopped reading codepoints.
+        size_t found = read_code_points(cp, start, end);
+        /// We need to start for the first time here, because first N - 1 codepoints mean nothing.
+        size_t i = N - 1;
+        size_t len = 0;
+        do
+        {
+            for (; i + N <= found; ++i)
+            {
+                UInt32 hash = 0;
+                for (size_t j = 0; j < N; ++j)
+                {
+                    hash <<= 8;
+                    hash += *(cp + i + j);
+                }
+                if (model[hash] == 0)
+                {
+                    model[hash] = 1;
+                    ++len;
+                }
+                ++model[hash];
```

Comment:
It will be `count  + 1`. Should not we set 0 initially?
Why value is float if we only count? :)